[Doc] k8s troubleshooting

kaisecheng opened the pull request:
Fixed: #14573 

Doc add troubleshooting session

robbavey commented:
@kaisecheng Thank you! I like the content you've put in here - I wonder if we can structure it to make the different issues and resolutions more contained togetherer/a little clearer. I put together something -  https://github.com/robbavey/logstash/blob/troubleshooting_thoughts/docsk8s/troubleshooting/ls-k8s-common-problems.asciidoc

WDYT?

kaisecheng commented:
It looks clearer 👍 and more actionable for users

kaisecheng commented:
@robbavey It will be great if you can push your changes to this one

robbavey commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -3,33 +3,49 @@
 
-Log2::
-Description.
+To view the log, `kubectl logs -f logstash-7477d46bb7-4lcnv`
```
Maybe we can also add how to set debug logging too?

robbavey commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -3,33 +3,49 @@
 You can use the standard Kubernetes `get` and `describe` commands to quickly gather details about any resources in your Logstash and Kubernetes environment.
 
-Examples...
+`kubectl get pod logstash-7477d46bb7-4lcnv`
```
Maybe show `kubectl get pods` along [these lines](https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-troubleshooting-methods.html#k8s-get-resources), then refer to the logstash pod mentioned in there?

robbavey commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -3,33 +3,49 @@
-Examples...
+`kubectl get pod logstash-7477d46bb7-4lcnv`
 
+`kubectl describe pod logstash-7477d46bb7-4lcnv`
```
Maybe add some sample output from `describe pod` and what a failing logstash pod might look like?

robbavey commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -3,33 +3,49 @@
+
+==== Heap Dump
+
+`jdk/bin/jcmd 1 GC.heap_dump /tmp/heap_dump.hprof`
```
We might also want to show how to extract the heap/thread dumps from the container - https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-jvm-heap-dumps.html#k8s_extracting_heap_dumps_from_the_elasticsearch_container

kaisecheng commented:
@robbavey I have updated the doc according to your suggestion. Please have a look.
- added sample output for commands
- added instruction of enabling debug log
- added extract file from container
- apply bash format to the scripts
- remove one layer in common problem session
- replace Logstash with {ls}

robbavey reviewed (commented):
Minor typo

robbavey reviewed (approved):
LGTM - can you move this into `needs doc review` for final approval please?

karenzone commented on docsk8s/troubleshooting/ls-k8s-common-problems.asciidoc:
```diff
@@ -1,25 +1,59 @@
+[[problem-delay]]
+==== {ls} startup process takes longer than `initialDelaySeconds`
+
+It may be necessary to review the time constraints of `readinessProbe` and `livenessProbe` to ensure that {ls} has enough time to start up and expose the health check endpoint for the `readiness` and `liveness` probes to access.
```
```suggestion
Review the time constraints of `readinessProbe` and `livenessProbe` to ensure that {ls} has enough time to start up and expose the health check endpoint for the `readiness` and `liveness` probes to access.
```

karenzone replied:
It's best when we speak directly to a user. In cases where we don't want to be too prescriptive, we can use "Consider <whatever it is they should try.>."

karenzone commented on docsk8s/troubleshooting/ls-k8s-common-problems.asciidoc:
```diff
@@ -1,25 +1,59 @@
+[[problem-insufficient]]
+==== Insufficient CPU or memory to start {ls}
+
+Review CPU and memory usage using `kubectl top pods` (note that this requires metrics server to be available for your Kubernetes implementation)
```
```suggestion
Review CPU and memory usage using `kubectl top pods` (requires metrics server to be available for your Kubernetes implementation).
```

karenzone commented on docsk8s/troubleshooting/ls-k8s-common-problems.asciidoc:
```diff
@@ -1,25 +1,59 @@
+Review CPU and memory usage using `kubectl top pods` (note that this requires metrics server to be available for your Kubernetes implementation)
+
+* Set the values of `cpu` and `memory` in your `Deployment` or `StatefulSet` appropriately.
+* Ensure that the JVM memory settings are set appropriately, note that the default `Xmx` value is `1g`, and we recommend that heap size is set to no more than 50-75% of total memory.
```
```suggestion
* Ensure that the JVM memory settings are set appropriately. The default `Xmx` value is `1g`, and we recommend that heap size is set to no more than 50-75% of total memory.
```

karenzone commented on docsk8s/troubleshooting/ls-k8s-common-problems.asciidoc:
```diff
@@ -1,25 +1,59 @@
+[float]
+[[problem-oom]]
+=== {ls} stops with OOM errors
+The status of {ls} shows `Ready` but the pod continues to stop running.
```
```suggestion
The status of {ls} shows `Ready,` but the pod repeatedly stops running.
```

karenzone commented on docsk8s/troubleshooting/ls-k8s-common-problems.asciidoc:
```diff
@@ -1,25 +1,59 @@
+=== {ls} stops with OOM errors
+The status of {ls} shows `Ready` but the pod continues to stop running.
+
+This can be caused by insufficient memory. If the memory of {ls} use more than the declared resource, kubernetes kill the pod immediately and the log of {ls} does not show any shutdown related message.
```
```suggestion
This situation can be caused by insufficient memory. If {ls} uses more memory than the declared resource, Kubernetes kills the pod immediately and the l{ls} log does not show any shutdown related message.
```

karenzone replied:
Kubernetes should be capitalized in every mention.

karenzone replied:
We try to use less violent language like "stopped" or "shut down" instead of "killed."  I see your "OOMKilled" message later in the docs, so I'm guessed that common language for K8s. Still, we should try to avoid when we can do so without confusing users.

kaisecheng replied:
we can use "Kubernetes shutdowns the pod" instead. "OOMKilled" is a pod status. "kill" is a common language in OS, eg. kill -9 process_id, to send a kill signal to terminate the process immediately

karenzone commented on docsk8s/troubleshooting/ls-k8s-common-problems.asciidoc:
```diff
@@ -1,25 +1,59 @@
+
+This can be caused by insufficient memory. If the memory of {ls} use more than the declared resource, kubernetes kill the pod immediately and the log of {ls} does not show any shutdown related message.
+
+Run `kubectl get event --watch` or `kubectl describe pod` to check kubernetes event if the status shows `OOMKilled`.
```
```suggestion
Run `kubectl get event --watch` or `kubectl describe pod` if the event status shows `OOMKilled`.
```

karenzone commented on docsk8s/troubleshooting/ls-k8s-common-problems.asciidoc:
```diff
@@ -1,25 +1,59 @@
+
+Run `kubectl get event --watch` or `kubectl describe pod` to check kubernetes event if the status shows `OOMKilled`.
+
+To resolve the problem, review JVM and memory settings as shown <<problem-insufficient, here>>
```
```suggestion
The resolution is similar to the remedy for the insufficient CPU or memory problem. 
Review JVM and memory settings as shown in <<problem-insufficient>>.
```

karenzone commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -1,35 +1,95 @@
 [[ls-k8s-troubleshooting-methods]]
 === Troubleshooting methods
```
```suggestion
=== Troubleshooting tips and suggestions
```

karenzone replied:
"Method" has a specific meaning in computing, so we should use different terminology when possible.

karenzone commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -1,35 +1,95 @@
 === Troubleshooting methods
 
-There are a number of methods that you can use to diagnose the state of your Logstash and Kubernetes system, both in the event of any problems, and as part of a day-to-day approach to ensuring that everything is running as expected.
+There are a number of methods that you can use to diagnose the state of your {ls} and Kubernetes system, both in the event of any problems, and as part of a day-to-day approach to ensuring that everything is running as expected.
```
```suggestion
Here are some approaches that you can use to diagnose the state of your {ls} and Kubernetes system, both in the event of any problems, and as part of a day-to-day approach to ensuring that everything is running as expected.
```

karenzone commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -1,35 +1,95 @@
+logstash-7477d46bb7-4lcnv               0/1     Pending       0                2m43s
+--
+
+If a Pod fails to reach the `Running` status after a few seconds, run this command to get more insights:
```
Nice.  The "run this command" construct is much better than using "the following."  Those are extra works that don't add any value, so it's nice that you left them out.

karenzone commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -1,35 +1,95 @@
-Following is a list of logs and locations that can be useful to monitor the health of your environment.
+{ls} Docker containers do not create log files by default. They log to standard output.
+
+To view the log, 
```
```suggestion
To view the log, run
```

karenzone commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -1,35 +1,95 @@
+[[ls-k8s-diagnostics]]
+=== Running diagnostics
+
+Sometimes, you may need to get thread dump and heap dump to debug hard problem. First you need to connect to the container, and then run the commands to take the diagnostics.
```
```suggestion
Thread dumps and heap dumps can be helpful when you are debugging hard problems. Connect to the container, and then run the commands to gather the diagnostics.
```

karenzone commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -1,35 +1,95 @@
+
+Sometimes, you may need to get thread dump and heap dump to debug hard problem. First you need to connect to the container, and then run the commands to take the diagnostics.
+
+==== Thread Dump
```
```suggestion
==== Thread dump
```

karenzone commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -1,35 +1,95 @@
+jdk/bin/jstack -l 1 > /tmp/jstack_output.txt
+--
+
+==== Heap Dump
```
```suggestion
==== Heap dump
```

karenzone reviewed (commented):
Nice work! Thank you. I left some comments for your consideration. Please let me know if you would like to discuss.

I used "mini-headings" like "Possible solutions" and "Work around" in the [Logstash Troubleshooting docs](https://www.elastic.co/guide/en/logstash/current/ts-logstash.html). Please keep those kinds of labels in mind if you think they might be helpful.

kaisecheng commented:
@karenzone Thank you so much for your suggestion and explanation. I have committed the changes and will keep in mind the mini heading.

karenzone commented:
@kaisecheng, thanks for all the changes. It's looking great!  
When you have time, let's discuss the changes you skipped so that I'll know if you disagree or if they were oversights. After that, I'll do the final review and signoff.

kaisecheng commented:
@karenzone I miss the suggestions that hide in the UI collapse 🤦. sorry

karenzone commented:
@kaise, it's all good. Github is "layered" and it's easy for things to hide.  That's why I asked instead of making assumptions that you chose to ignore the suggestions.  :-) As you know, you are _never_ obligated to accept suggestions, and we can discuss anything that you don't like or doesn't make sense to reach a resolution that makes us both happy.

karenzone commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -1,35 +1,95 @@
+  Warning  FailedScheduling  34s (x2 over 115s)  default-scheduler  0/1 nodes are available: 1 Insufficient cpu.
+--
+
+You can check the CPU and memory resources by running this command
```
Needs colon

karenzone commented on docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc:
```diff
@@ -1,35 +1,95 @@
-Following is a list of logs and locations that can be useful to monitor the health of your environment.
+{ls} Docker containers do not create log files by default. They log to standard output.
+
+To view the log, run
```
Needs colon.

karenzone reviewed (approved):
Nicely done, @kaisecheng. Thank you!
LGTM
